File: kutuphane-rezervasyon-frontend/gatling-tests/src/test/java/simulations/StressTest.java
package simulations;

import io.gatling.javaapi.core.*;
import io.gatling.javaapi.http.*;

import java.time.Duration;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

public class StressTest extends Simulation {

    HttpProtocolBuilder httpProtocol = http
        .baseUrl("http://localhost:8080")
        .acceptHeader("application/json")
        .contentTypeHeader("application/json")
        .header("Origin", "http://localhost:4201");

    // Stres testi - Sistemin limitlerini test eder
    ScenarioBuilder stressScenario = scenario("Stress Test - High Load")
        .exec(http("Login")
            .post("/api/auth/login")
            .body(StringBody("{\"email\":\"root@kutuphane.com\",\"password\":\"root123\"}"))
            .asJson()
            .check(status().is(200))
            .check(jsonPath("$.userId").exists()))
        .pause(1)
        .repeat(10).on(
            exec(http("Rapid Room Check")
                .get("/api/rooms/available")
                .queryParam("date", "2025-11-20")
                .queryParam("timeSlotId", "1")
                .check(status().in(200, 401, 429, 500))) // 429: Too Many Requests, 500: Server Error
            .pause(Duration.ofMillis(100))
        );

    {
        setUp(
            stressScenario.injectOpen(
                // Kademeli olarak kullanıcı sayısını artır
                rampUsers(10).during(Duration.ofSeconds(10)),
                constantUsersPerSec(5).during(Duration.ofSeconds(20)),
                rampUsers(50).during(Duration.ofSeconds(30)),
                constantUsersPerSec(20).during(Duration.ofSeconds(40))
            )
        ).protocols(httpProtocol)
         .maxDuration(Duration.ofMinutes(2))
         .assertions(
             global().responseTime().max().lt(10000),
             global().successfulRequests().percent().gt(80.0)
         );
    }
}

File: kutuphane-rezervasyon-frontend/gatling-tests/src/test/scala/simulations/StressTest.scala
package simulations

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

class StressTest extends Simulation {

  val baseUrl = "http://localhost:8080"
  
  val httpProtocol = http
    .baseUrl(baseUrl)
    .acceptHeader("application/json")
    .contentTypeHeader("application/json")

  // Stres testi - Sistemin limitlerini test eder
  val stressScenario = scenario("Stress Test - High Load")
    .exec(http("Login")
      .post("/api/auth/login")
      .body(StringBody("""{"email":"user@example.com","password":"user123"}""")).asJson
      .check(status.is(200))
      .check(jsonPath("$.token").saveAs("authToken")))
    .pause(1)
    .repeat(10) {
      exec(http("Rapid Room Check")
        .get("/api/rooms/available")
        .queryParam("date", "2025-11-20")
        .queryParam("timeSlotId", "1")
        .header("Authorization", "Bearer ${authToken}")
        .check(status.in(200, 429, 500))) // 429: Too Many Requests, 500: Server Error
      .pause(100.milliseconds)
    }

  setUp(
    stressScenario.inject(
      // Kademeli olarak kullanıcı sayısını artır
      rampUsers(10).during(10.seconds),
      constantUsersPerSec(5).during(20.seconds),
      rampUsers(50).during(30.seconds),
      constantUsersPerSec(20).during(40.seconds)
    )
  ).protocols(httpProtocol)
   .maxDuration(2.minutes)
   .assertions(
     global.responseTime.max.lt(10000),
     global.successfulRequests.percent.gt(80)
   )
}
